connect/runtime/src/main/java/org/apache/kafka/connect/util/ConnectUtils.java

```diff
@@ -72,6 +74,45 @@ static String lookupKafkaClusterId(Admin adminClient) {
         }
     }
 
+    /**
+     * Log a warning when the user attempts to override a property that cannot be overridden.
+     * @param props the configuration properties provided by the user
+     * @param key the name of the property to check on
+     * @param expectedValue the expected value for the property
+     * @param justification the reason the property cannot be overridden.
+     *                      Will follow the phrase "The value... for the... property will be ignored as it cannot be overridden ".
+     *                      For example, one might supply the message "in connectors with the DLQ feature enabled" for this parameter.
+     * @param caseSensitive whether the value should match case-insensitively
+     */
+    public static void warnOnOverriddenProperty(
+            Map<String, ?> props,
+            String key,
+            String expectedValue,
+            String justification,
+            boolean caseSensitive) {
+        overriddenPropertyWarning(props, key, expectedValue, justification, caseSensitive).ifPresent(log::warn);
+    }
+
+    // Visible for testing
+    static Optional<String> overriddenPropertyWarning(
+            Map<String, ?> props,
+            String key,
+            String expectedValue,
+            String justification,
+            boolean caseSensitive) {
+        Predicate<String> matchesExpectedValue = caseSensitive ? expectedValue::equals : expectedValue::equalsIgnoreCase;
+        String value = Optional.ofNullable(props.get(key)).map(Object::toString).orElse(null);
+        if (value != null && !matchesExpectedValue.test(value)) {
+            return Optional.of(String.format(
+                    "The value '%s' for the '%s' property will be ignored as it cannot be overridden%s. "
+                            + "The value '%s' will be used instead.",
```

Comment:
I suppose not! There are two classes that currently leverage this utility method: `Worker` and `DistributedConfig`.

In the former, each invocation of `warnOnOverriddenProperty` is followed closely (if not immediately) by a line that overrides any user-supplied values that don't match the expected value.

However, in `DistributedConfig`, warnings are emitted for values that are ignored and overridden when constructing Kafka clients for some but not all `KafkaBasedLog` instances: specifically, the ones used for the config and offset backing stores. It's possible that a user might specify `enable.idempotence=false` in their worker config with the intention of disabling idempotent writes for the producers used for the status backing store (since the default for this property was updated from `false` to `true` in [KIP-679](https://cwiki.apache.org/confluence/display/KAFKA/KIP-679%3A+Producer+will+enable+the+strongest+delivery+guarantee+by+default)). With the current behavior for this PR, that change would have an effect (and the values logged for the producer for the status backing store would confirm this), but the warning message would still be logged, which might lead to some confusion.

At the bare minimum we can and should update this message to be more specific about where exactly these values will be ignored, but in addition to that, I've taken a stab at restructuring the reusable utility function to actually perform the override in addition to just warning about it, like you suggested in a comment below. This makes it harder to log warnings about overrides without actually ensuring that the overrides have taken place.